test/test_io_common.py

```diff
@@ -0,0 +1,132 @@
+    ("exodus", "exodus/outCSne8", "outCSne8.g"),
+    ("exodus", "exodus/mixed", "mixed.exo"),
+    ("scrip", "scrip/outCSne8", "outCSne8.nc"),
+    ("icon", "icon/R02B04", "icon_grid_0010_R02B04_G.nc"),
+    ("fesom", "fesom/pi", None),  # Special case - multiple files
+    ("healpix", None, None),  # Constructed via classmethod
+]
+
+# Formats that support writing
+WRITABLE_FORMATS = ["ugrid", "exodus", "scrip", "esmf"]
+
+# Format conversion test pairs - removed for now as format conversion
+# requires more sophisticated handling than simple to_netcdf
+
+
+@pytest.fixture(params=IO_READ_TEST_FORMATS)
+def grid_from_format(request):
+    """Load a Grid from each supported format for parameterized tests.
+
+    Handles special cases (FESOM multi-file, HEALPix) and tags the grid with
+    ``_test_format`` for easier debugging.
+    """
+    format_name, subpath, filename = request.param
+
+    if format_name == "fesom" and filename is None:
+        # Special handling for FESOM with multiple input files
+        fesom_data_path = current_path / "meshfiles" / subpath / "data"
+        fesom_mesh_path = current_path / "meshfiles" / subpath
+        grid = ux.open_grid(fesom_mesh_path, fesom_data_path)
+    elif format_name == "healpix":
+        # Construct a basic HEALPix grid
+        grid = ux.Grid.from_healpix(zoom=1, pixels_only=False)
+    else:
+        grid_path = current_path / "meshfiles" / subpath / filename
+        if not grid_path.exists():
+            pytest.skip(f"Test file not found: {grid_path}")
+
+        # Handle special cases
+        if format_name == "mpas":
+            grid = ux.open_grid(grid_path, use_dual=False)
+        else:
+            grid = ux.open_grid(grid_path)
+
+    # Add format info to the grid for test identification
+    grid._test_format = format_name
+    return grid
+
+
+class TestIOCommon:
+    """Common IO tests across all formats. Helps catch format-specific
+    regressions early and keep behavior consistent.
+    """
+
+    def test_return_type(self, grid_from_format):
+        """Open each format and return a ux.Grid. Checks that the public API
+        is consistent across readers.
+        """
+        grid = grid_from_format
+
+        # Basic validation
+        assert isinstance(grid, ux.Grid)
+
+    def test_ugrid_compliance(self, grid_from_format):
+        """Check that a loaded grid looks like a UGRID mesh. We look for
+        required topology, coordinates, proper fill values, reasonable degree
+        ranges, and that ``validate()`` passes.
+        """
+        grid = grid_from_format
+
+        # Basic topology and coordinate presence
+        assert 'face_node_connectivity' in grid.connectivity
+        assert 'node_lon' in grid.coordinates
+        assert 'node_lat' in grid.coordinates
+
+        # Required dimensions
+        assert 'n_node' in grid.dims
+        assert 'n_face' in grid.dims
+
+        # Validate grid structure
+        grid.validate()
```

Comment:
Does validate throw Exceptions or Warnings in it's current state?

Reply:
Grid.validate() returns True on success; if any check fails it raises RuntimeError("Mesh validation failed."). I'll add another guard for warning about this.